stl/inc/algorithm

```diff
@@ -3899,6 +3999,118 @@ _CONSTEXPR20 void pop_heap(_RanIt _First, _RanIt _Last) {
     _STD pop_heap(_First, _Last, less<>());
 }
 
+#ifdef __cpp_lib_concepts
+namespace ranges {
+    // VARIABLE ranges::pop_heap
+    template <class _It, class _Pr, class _Pj>
+    constexpr void _Pop_heap_hole_by_index(_It _First, iter_difference_t<_It> _Hole,
+        const iter_difference_t<_It> _Bottom, iter_value_t<_It>&& _Val, _Pr _Pred, _Pj _Proj) {
+        // percolate _Hole to _Bottom, then push _Val, using _Pred
+        _STL_INTERNAL_STATIC_ASSERT(random_access_iterator<_It>);
+        _STL_INTERNAL_STATIC_ASSERT(sortable<_It, _Pr, _Pj>);
+        _STL_INTERNAL_CHECK(_Bottom != 0);
+        using _Diff      = iter_difference_t<_It>;
+        const _Diff _Top = _Hole;
+        _Diff _Idx       = _Hole;
+
+        // Check whether _Idx can have a child before calculating that child's index, since
+        // calculating the child's index can trigger integer overflows
+        const _Diff _Max_sequence_non_leaf = (_Bottom - 1) >> 1; // shift for codegen

```

Comment:
Is that shift thingy still needed?

Mat would be sad

Reply:
Division by 2^n is equivalent to shifting right by n for unsigned integral types and non-negative values of signed integral types (https://godbolt.org/z/R-VzUb). Compilers aren't smart enough to determine that all of these differences that we use to index into ranges are non-negative. (Or, for the most part, to listen when we tell them so.)